### PR TITLE
Add more types and documentation to Mock

### PR DESCRIPTION
**Pull request by matanlurey**
Mock is sort of a black-box of magic, and adding some documentation hopefully will help contributors in the future as well as give our users a better idea of what exactly they are using.

A tiny bit of refactoring on top of that, but no breaking API changes.

I've upped the pubspec to `+1`, but I can revert that if you'd prefer I don't.

/cc @srawlins

### CODE REVIEW

File: test/mockito_test.dart [hunk 1]
```diff
@@ -33,13 +33,9 @@ abstract class AbstractFoo implements Foo {
   String baz();
 }
 
-class MockFoo extends AbstractFoo with Mock {
-  noSuchMethod(i) => super.noSuchMethod(i);
```
Comment (srawlins) [inline test/mockito_test.dart]: omg thank you; been meaning to do this forever.

**Comment by srawlins** [comment]
LGTM; CC @TedSander

File: test/mockito_test.dart (hunk 1 above)
Comment (matanlurey) [inline test/mockito_test.dart]: 🍔

**Comment by TedSander** [comment]
LGTM

**Pull request merged**